metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java

```diff
@@ -0,0 +1,504 @@
+        preVersionChangeImage.write(writer, new ImageWriterOptions.Builder().
+                        setMetadataVersion(change.newVersion()).
+                        setLossHandler(loss -> {
+                            log.warn("{}", loss.getMessage());
+                        }).
+                        build());
+    }
+
+    @Override
+    public void handleSnapshot(SnapshotReader<ApiMessageAndVersion> reader) {
+        eventQueue.append(() -> {
+            try {
+                MetadataDelta delta = new MetadataDelta.Builder().
+                        setImage(image).
+                        build();
+                SnapshotManifest manifest = loadSnapshot(delta, reader);
+                try {
+                    image = delta.apply(manifest.provenance());
+                } catch (Throwable e) {
+                    faultHandler.handleFault("Error generating new metadata image from " +
+                            "snapshot at offset " + reader.lastContainedLogOffset(), e);
+                    return;
+                }
+                for (MetadataPublisher publisher : publishers) {
+                    try {
+                        publisher.publishSnapshot(delta, image, manifest);
+                    } catch (Throwable e) {
+                        faultHandler.handleFault("Unhandled error publishing the new metadata " +
+                                "image from snapshot at offset " + reader.lastContainedLogOffset() +
+                                    " with publisher " + publisher.name(), e);
+                    }
+                }
+                metrics.updateLastAppliedImageProvenance(image.provenance());
+            } catch (Throwable e) {
+                // This is a general catch-all block where we don't expect to end up;
+                // failure-prone operations should have individual try/catch blocks around them.
+                faultHandler.handleFault("Unhandled fault in MetadataLoader#handleSnapshot. " +
+                        "Snapshot offset was " + reader.lastContainedLogOffset(), e);
+            } finally {
+                reader.close();
+            }
+        });
+    }
+
+    /**
+     * Load a snapshot. This is relatively straightforward since we don't track as many things as
+     * we do in loadLogDelta. Additionally, it is an error for a version change to occur unless
+     * it is the very first record. The main complication here is that we have to maintain an index
+     * of what record we are processing so that we can give useful error messages.
+     *
+     * @param delta     The metadata delta we are preparing.
+     * @param reader    The reader which yields the snapshot batches.
+     * @return          A manifest of what was loaded.
+     */
+    SnapshotManifest loadSnapshot(
+            MetadataDelta delta,
+            SnapshotReader<ApiMessageAndVersion> reader
+    ) {
+        long startNs = time.nanoseconds();
+        int snapshotIndex = 0;
```

Comment:
Not sure how useful this index is given the presence of control records in the snapshot. Why not use the offset from `Batch` as we do when processing log records?

Reply:
Hmm. I'm actually a bit confused about what that field in the batch contains, since from the KRaft point of view all the records in a snapshot are at the same "offset"

Since snapshotIndex is ported from the old code, perhaps we can look into it a bit later? Or do a follow-on JIRA...